File: ios/LaneShadow/Sandbox/Stories/LSInputStories.swift
import LaneShadowTheme
import NativeSandbox
import SwiftUI

@MainActor
enum LSInputStories {
    static let all: [Story] = [
        Story(
            id: "atoms.textfield.default",
            tier: .atom,
            component: "LSTextField",
            name: "Default",
            summary: "Token-backed single-line field with placeholder text."
        ) { _ in
            TextFieldDefaultStory()
        },
        Story(
            id: "atoms.textfield.focused",
            tier: .atom,
            component: "LSTextField",
            name: "Focused",
            summary: "Focused field with copper border treatment."
        ) { _ in
            TextFieldFocusedStory()
        },
        Story(
            id: "atoms.textfield.error",
            tier: .atom,
            component: "LSTextField",
            name: "Error",
            summary: "Error treatment with helper copy."
        ) { _ in
            TextFieldErrorStory()
        },
        Story(
            id: "atoms.textfield.disabled",
            tier: .atom,
            component: "LSTextField",
            name: "Disabled",
            summary: "Disabled field with suppressed editing."
        ) { _ in
            TextFieldDisabledStory()
        },
        Story(
            id: "atoms.textfield.with-icon",
            tier: .atom,
            component: "LSTextField",
            name: "With Icon",
            summary: "Leading icon slot routed through LSIcon."
        ) { _ in
            TextFieldIconStory()
        },
        Story(
            id: "atoms.textarea.default",
            tier: .atom,
            component: "LSTextArea",
            name: "Default",
            summary: "Token-backed multi-line field."
        ) { _ in
            TextAreaDefaultStory()
        },
        Story(
            id: "atoms.textarea.autogrow",
            tier: .atom,
            component: "LSTextArea",
            name: "Auto Grow",
            summary: "Auto-growing multi-line field sized from content."
        ) { _ in
            TextAreaAutogrowStory()
        },
    ]
}

private struct TextFieldDefaultStory: View {
    @State private var value = ""

    var body: some View {
        InputStoryFrame {
            LSTextField(value: $value, placeholder: "Email")
        }
    }
}

private struct TextFieldFocusedStory: View {
    @State private var value = "Marin Headlands"

    var body: some View {
        InputStoryFrame {
            LSTextField(value: $value, placeholder: "Destination", state: .focused)
        }
    }
}

private struct TextFieldErrorStory: View {
    @State private var value = "xxxxxx"

    var body: some View {
        InputStoryFrame {
            LSTextField(
                value: $value,
                placeholder: "Destination",
                state: .error,
                helperText: "Location not recognized. Try a city or landmark."
            )
        }
    }
}

private struct TextFieldDisabledStory: View {
    @State private var value = "Set by Navigator"

    var body: some View {
        InputStoryFrame {
            LSTextField(
                value: $value,
                placeholder: "Set automatically",
                state: .disabled,
                helperText: "This field is managed upstream."
            )
        }
    }
}

private struct TextFieldIconStory: View {
    @State private var value = ""

    var body: some View {
        InputStoryFrame {
            LSTextField(
                value: $value,
                placeholder: "Search routes",
                leadingIcon: .route
            )
        }
    }
}

private struct TextAreaDefaultStory: View {
    @State private var value = ""

    var body: some View {
        InputStoryFrame {
            LSTextArea(value: $value, placeholder: "Add notes", autoGrow: false)
        }
    }
}

private struct TextAreaAutogrowStory: View {
    @State private var value = """
    Scenic turnout before the ridge.
    Coffee stop near the lake.
    Fuel before the canyon stretch.
    """

    var body: some View {
        InputStoryFrame {
            LSTextArea(value: $value, placeholder: "Route notes", autoGrow: true)
        }
    }
}

private struct InputStoryFrame<Content: View>: View {
    @Environment(\.theme) private var theme

    private let content: Content

    init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: theme.space.lg) {
            content
        }
        .padding(theme.space.lg)
        .background(theme.colors.surface.default)
    }
}

File: android/app/src/debug/java/com/laneshadow/sandbox/stories/LSInputStories.kt
package com.laneshadow.sandbox.stories

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.laneshadow.theme.LaneShadowTheme
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.atoms.InputState
import com.laneshadow.ui.atoms.LSTextArea
import com.laneshadow.ui.atoms.LSTextField
import com.nativesandbox.model.ComponentTier
import com.nativesandbox.model.Story

object LSInputStories {
    val all: List<Story> = listOf(
        inputStory(
            id = "atoms.textfield.default",
            component = "LSTextField",
            name = "Default",
            summary = "Single-line input in the default subtle-border state.",
            state = InputState.Default,
            multiline = false,
            initialValue = "",
            placeholder = "Enter destination",
        ),
        inputStory(
            id = "atoms.textfield.focused",
            component = "LSTextField",
            name = "Focused",
            summary = "Single-line input with the focused border token.",
            state = InputState.Focused,
            multiline = false,
            initialValue = "Marin Headlands",
            placeholder = "Enter destination",
        ),
        inputStory(
            id = "atoms.textfield.error",
            component = "LSTextField",
            name = "Error",
            summary = "Single-line input using the danger border token.",
            state = InputState.Error,
            multiline = false,
            initialValue = "???",
            placeholder = "Enter destination",
        ),
        inputStory(
            id = "atoms.textfield.disabled",
            component = "LSTextField",
            name = "Disabled",
            summary = "Single-line input rendered in a non-interactive state.",
            state = InputState.Disabled,
            multiline = false,
            initialValue = "Set automatically",
            placeholder = "Enter destination",
        ),
        inputStory(
            id = "atoms.textarea.default",
            component = "LSTextArea",
            name = "Default",
            summary = "Multi-line input in the default subtle-border state.",
            state = InputState.Default,
            multiline = true,
            initialValue = "",
            placeholder = "Add ride notes",
        ),
        inputStory(
            id = "atoms.textarea.focused",
            component = "LSTextArea",
            name = "Focused",
            summary = "Multi-line input with the focused border token.",
            state = InputState.Focused,
            multiline = true,
            initialValue = "Scenic route via the ridge road.",
            placeholder = "Add ride notes",
        ),
        inputStory(
            id = "atoms.textarea.error",
            component = "LSTextArea",
            name = "Error",
            summary = "Multi-line input using the danger border token.",
            state = InputState.Error,
            multiline = true,
            initialValue = "Need a clearer meeting point.",
            placeholder = "Add ride notes",
        ),
        inputStory(
            id = "atoms.textarea.disabled",
            component = "LSTextArea",
            name = "Disabled",
            summary = "Multi-line input rendered in a non-interactive state.",
            state = InputState.Disabled,
            multiline = true,
            initialValue = "Navigator will generate a route summary.",
            placeholder = "Add ride notes",
        ),
    )
}

private fun inputStory(
    id: String,
    component: String,
    name: String,
    summary: String,
    state: InputState,
    multiline: Boolean,
    initialValue: String,
    placeholder: String,
): Story =
    Story(
        id = id,
        tier = ComponentTier.Atom,
        component = component,
        name = name,
        summary = summary,
        content = {
            LaneShadowTheme {
                var text by remember { mutableStateOf(initialValue) }
                InputStoryColumn {
                    if (multiline) {
                        LSTextArea(
                            value = text,
                            onValueChange = { text = it },
                            state = state,
                            placeholder = placeholder,
                            maxRows = 6,
                        )
                    } else {
                        LSTextField(
                            value = text,
                            onValueChange = { text = it },
                            state = state,
                            placeholder = placeholder,
                        )
                    }
                }
            }
        },
    )

@Composable
private fun InputStoryColumn(content: @Composable () -> Unit) {
    val theme = LocalLaneShadowTheme.current

    Column(
        modifier = Modifier.padding(theme.space.lg),
        verticalArrangement = Arrangement.spacedBy(theme.space.md),
    ) {
        content()
    }
}
